torch/_higher_order_ops/while_loop.py

```diff
@@ -0,0 +1,126 @@
+import torch.utils._pytree as pytree
+
+from torch._C import DispatchKey
+from torch._higher_order_ops.utils import (
+    _has_potential_branch_input_alias,
+    _has_potential_branch_input_mutation,
+    _maybe_run_with_interpreter,
+    autograd_not_implemented,
+    UnsupportedAliasMutationException,
+)
+from torch._ops import HigherOrderOperator
+
+from torch._subclasses.fake_tensor import FakeTensorMode
+from torch.fx.experimental.proxy_tensor import (
+    disable_proxy_modes_tracing,
+    make_fx,
+    ProxyTorchDispatchMode,
+    track_tensor_tree,
+)
+
+
+while_loop_op = HigherOrderOperator("while_loop")
+
+
+def while_loop(cond_fn, body_fn, operands):
+    # TODO(yidi): turn on dynamo in eager-mode
+    assert isinstance(operands, (tuple, list))
+    return while_loop_op(cond_fn, body_fn, operands)
+
+
+@while_loop_op.py_impl(DispatchKey.CompositeExplicitAutograd)
+def while_loop_dense(cond_fn, body_fn, operands):
+    init_val = operands
+    while cond_fn(*init_val):
```

Comment:
should probably check that cond_fn is a bool or a one-elemnt BoolTensor?

Reply:
Yeah, I can add the check. The cond_fn may have to return a boolean Scalar tensor because dynamo will specialize the python bool to a constant when we speculate cond_fn.

Reply:
Dynamo's not going to run on this function, right? So we can easily add the check here.